src/network/core/http_curl.cpp

```diff
@@ -12,216 +12,165 @@
+		curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 5L);
 
-	if (_http_requests.empty()) return;
+		/* Give the connection about 10 seconds to complete. */
+		curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 10L);
 
-	for (auto it = _http_requests.begin(); it != _http_requests.end(); /* nothing */) {
-		NetworkHTTPRequest *cur = *it;
+		/* Set a buffer of 100KiB, as the default of 16KiB seems a bit small. */
+		curl_easy_setopt(curl, CURLOPT_BUFFERSIZE, 100L * 1024L);
 
-		if (cur->Receive()) {
-			it = _http_requests.erase(it);
-			delete cur;
-			continue;
-		}
+		/* Fail our call if we don't receive a 2XX return value. */
+		curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1L);
 
-		++it;
+		/* Prepare POST body and URI. */
+		if (request->data != nullptr) {
+			curl_easy_setopt(curl, CURLOPT_POST, 1L);
+			curl_easy_setopt(curl, CURLOPT_POSTFIELDS, request->data);
+		}
+		curl_easy_setopt(curl, CURLOPT_URL, request->uri.c_str());
+
+		/* Setup our (C-style) callback function which we pipe back into the callback. */
+		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, +[](char *ptr, size_t size, size_t nmemb, void *userdata) -> size_t {
+			Debug(net, 4, "HTTP callback: {} bytes", size * nmemb);
+			HTTPCallback *callback = static_cast<HTTPCallback *>(userdata);
+			callback->OnReceiveData(ptr, size * nmemb);
+			return size * nmemb;
+		});
+		curl_easy_setopt(curl, CURLOPT_WRITEDATA, request->callback);
+
+		/* Create a callback from which we can cancel. Sadly, there is no other
+		 * thread-safe way to do this. If the connection went idle, it can take
+		 * up to a second before this callback is called. There is little we can
+		 * do about this. */
+		curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 0L);
+		curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, +[](void *userdata, curl_off_t dltotal, curl_off_t dlnow, curl_off_t ultotal, curl_off_t ulnow) -> int {
+			return _http_thread_exit ? 1 : 0;
+		});
+
+		/* Perform the request. */
+		CURLcode res = curl_easy_perform(curl);
+
+		if (res == CURLE_OK) {
+			Debug(net, 1, "HTTP request succeeded");
+			request->callback->OnReceiveData(nullptr, 0);
+		} else {
+			Debug(net, 0, "HTTP request failed: {}", curl_easy_strerror(res));
+			request->callback->OnFailure();
+		}
 	}
+
+	curl_easy_cleanup(curl);
 }
 
 void NetworkHTTPInitialize()
 {
 	curl_global_init(CURL_GLOBAL_DEFAULT);
 
-	/* Create a share that tracks DNS, SSL session, and connections. As this
-	 * always runs in the same thread, sharing a connection should be fine. */
-	_curl_share = curl_share_init();
-	curl_share_setopt(_curl_share, CURLSHOPT_SHARE, CURL_LOCK_DATA_DNS);
-	curl_share_setopt(_curl_share, CURLSHOPT_SHARE, CURL_LOCK_DATA_SSL_SESSION);
-	curl_share_setopt(_curl_share, CURLSHOPT_SHARE, CURL_LOCK_DATA_CONNECT);
+	_http_thread_exit = false;
+	StartNewThread(&_http_thread, "ottd:http", &HttpThread);
 }
 
 void NetworkHTTPUninitialize()
 {
-	curl_share_cleanup(_curl_share);
 	curl_global_cleanup();
+
+	_http_thread_exit = true;
```

Comment:
"_http_thread_exit = true" + "notify_one" must be inside the mutex.
Otherwise they could be executed just before the http-thread goes to sleep.

On the plus side, "_http_thread_exit" is then protected by the mutex, and no longer needs to be atomic.

Reply:
Hmm, not sure I completely agree.

That is, for the `notify_one`, you are absolutely right. I should get the lock before I notify :) Nice catch there :D
But for `_http_thread_exit`, the lock is not meant to guard that; at least, I wrote the lock to guard `_http_requests`. Then I think I get where you are coming for, but I also think you missed the read in `CURLOPT_XFERINFOFUNCTION`. As this is also outside the lock.

Now I could add a lock in that callback, but that seems wasteful. It is just easier to make `_http_thread_exit` atomic. Although I doubt it matters for a `bool`, but that is yet-another-story :P

Now we could also add a signal on the queue to notify: job's done, but to prepare for a functional "cancel", I needed some logic in `CURLOPT_XFERINFOFUNCTION` anyway, and this felt cheaper.

At least, that is my reasoning .. now you will point to where my reasoning failed :D